NPM: Authentication with private registry fails with ERR_INVALID_AUTH

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

npm

### Package manager version

9.5.1

### Language version

_No response_

### Manifest location and content before the Dependabot update

_No response_

### dependabot.yml content

version: 2
registries:
    npm-internal:
        type: npm-registry
        url: pkgs.dev.azure.com/org/project/_packaging/registry/npm/registry
        token: ${{NPM_PAT}}
        replaces-base: true
updates:
    - package-ecosystem: 'npm'
      directory: '/'
      registries:
          - npm-internal
      schedule:
          interval: 'weekly'


### Updated dependency

_No response_

### What you expected to see, versus what you actually saw

This problem started happening when using NPM version 8 or newer.

In our configuration, we're using Azure Artifacts as the _only_ registry, with upstream feeds to npmjs et.al.
Our .npmrc looks like:
```
registry=https://pkgs.dev.azure.com/org/project/_packaging/registry/npm/registry/

always-auth=true
```
This also means that every package in `package-lock.json` will have the above registry as the `resolved` URL.
From what I can deduce by looking at the npmrc_builder.rb file in dependabot-core, this means that this registry will be classified as a `global` registry, see: https://github.com/dependabot/dependabot-core/blob/f9754d41004cb5a507b3cd4920fbf48551f9aced/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb#L84

This in turn leads to this part of the code https://github.com/dependabot/dependabot-core/blob/f9754d41004cb5a507b3cd4920fbf48551f9aced/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb#L165 creating the following `.npmrc` :

```
registry=https://pkgs.dev.azure.com/org/project/_packaging/registry/npm/registry/
_authToken=token
always-auth=true
```

This is a problem, because since NPM v8, all auth-related values need to be scoped, see: https://docs.npmjs.com/cli/v9/configuring-npm/npmrc?v=true#auth-related-configuration

A correct `.npmrc` needs to look like:

```
registry=https://pkgs.dev.azure.com/org/project/_packaging/registry/npm/registry/
//pkgs.dev.azure.com/org/project/_packaging/registry/npm/registry/:_authToken=token
always-auth=true
```




### Native package manager behavior

_No response_

### Images of the diff or a link to the PR, issue, or logs

_No response_

### Smallest manifest that reproduces the issue

_No response_

I get those logs from our CI:

> npm ERR! code ERR_INVALID_AUTH
npm ERR! Invalid auth configuration found: `_auth` must be renamed to `//pkgs.dev.azure.com/mycompany/_packaging/mycompany/npm/registry/:_auth` in project config
npm ERR! Please run `npm config fix` to repair your configuration.`

The content of the .npmrc file looks like:
```
registry=https://pkgs.dev.azure.com/mycompany/_packaging/mycompany/npm/registry/
always-auth=true
```

Does that help you?

--- a/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb
+++ b/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb
@@ -32,9 +32,17 @@ module Dependabot
               build_npmrc_content_from_lockfile
             end
 
-          return initial_content || "" unless registry_credentials.any?
+          final_content = initial_content || ""
 
-          ([initial_content] + credential_lines_for_npmrc).compact.join("\n")
+          return final_content unless registry_credentials.any?
+
+          credential_lines_for_npmrc.each do |credential_line|
+            next if final_content.include?(credential_line)
+
+            final_content = [final_content, credential_line].reject(&:empty?).join("\n")
+          end
+
+          final_content
         end
 
         # PROXY WORK

--- a/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb
+++ b/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb
@@ -113,7 +113,7 @@ module Dependabot
           token = global_registry.fetch("token", nil)
           return "" unless token
 
-          auth_line(token) + "\n"
+          auth_line(token, global_registry.fetch("registry")) + "\n"
         end
 
         def yarnrc_global_registry_auth_line

--- a/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb
+++ b/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npmrc_builder.rb
@@ -105,15 +105,7 @@ module Dependabot
           token = global_registry.fetch("token", nil)
           return "" unless token
 
-          if token.include?(":")
-            encoded_token = Base64.encode64(token).delete("\n")
-            "_auth = #{encoded_token}\n"
-          elsif Base64.decode64(token).ascii_only? &&
-                Base64.decode64(token).include?(":")
-            "_auth = #{token.delete("\n")}\n"
-          else
-            "_authToken = #{token}\n"
-          end
+          auth_line(token) + "\n"
         end
 
         def yarnrc_global_registry_auth_line
@@ -122,12 +114,12 @@ module Dependabot
 
           if token.include?(":")
             encoded_token = Base64.encode64(token).delete("\n")
-            "npmAuthIdent: \"#{encoded_token}\"\n"
+            "npmAuthIdent: \"#{encoded_token}\""
           elsif Base64.decode64(token).ascii_only? &&
                 Base64.decode64(token).include?(":")
-            "npmAuthIdent: \"#{token.delete("\n")}\"\n"
+            "npmAuthIdent: \"#{token.delete("\n")}\""
           else
-            "npmAuthToken: \"#{token}\"\n"
+            "npmAuthToken: \"#{token}\""
           end
         end
 
@@ -230,24 +222,33 @@ module Dependabot
             token = cred.fetch("token", nil)
             next unless token
 
-            # We need to ensure the registry uri ends with a trailing slash in the npmrc file
-            # but we do not want to add one if it already exists
-            registry_with_trailing_slash = registry.sub(%r{\/?$}, "/")
-            if token.include?(":")
-              encoded_token = Base64.encode64(token).delete("\n")
-              lines << "//#{registry_with_trailing_slash}:_auth=#{encoded_token}"
-            elsif Base64.decode64(token).ascii_only? &&
-                  Base64.decode64(token).include?(":")
-              lines << %(//#{registry_with_trailing_slash}:_auth=#{token.delete("\n")})
-            else
-              lines << "//#{registry_with_trailing_slash}:_authToken=#{token}"
-            end
+            lines << auth_line(token, registry)
           end
 
           return lines unless lines.any? { |str| str.include?("auth=") }
 
           # Work around a suspected yarn bug
           ["always-auth = true"] + lines
+        end
+
+        def auth_line(token, registry = nil)
+          auth = if token.include?(":")
+                   encoded_token = Base64.encode64(token).delete("\n")
+                   "_auth=#{encoded_token}"
+                 elsif Base64.decode64(token).ascii_only? &&
+                       Base64.decode64(token).include?(":")
+                   "_auth=#{token.delete("\n")}"
+                 else
+                   "_authToken=#{token}"
+                 end
+
+          return auth unless registry
+
+          # We need to ensure the registry uri ends with a trailing slash in the npmrc file
+          # but we do not want to add one if it already exists
+          registry_with_trailing_slash = registry.sub(%r{\/?$}, "/")
+
+          "//#{registry_with_trailing_slash}:#{auth}"
         end
 
         def npmrc_scoped_registries

--- a/npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npmrc_builder_spec.rb
+++ b/npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npmrc_builder_spec.rb
@@ -332,9 +332,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
           it "adds a global registry line, and auth details" do
             expect(npmrc_content)
               .to eq("registry = https://npm.fury.io/dependabot\n" \
-                     "_authToken=my_token\n" \
-                     "always-auth = true\n" \
-                     "//npm.fury.io/dependabot/:_authToken=my_token")
+                     "//npm.fury.io/dependabot/:_authToken=my_token\n" \
+                     "always-auth = true")
           end
 
           context "and an npmrc file" do
@@ -346,9 +345,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                        "strict-ssl = true\n" \
                        "//npm.fury.io/dependabot/:_authToken=secret_token\n" \
                        "registry = https://npm.fury.io/dependabot\n" \
-                       "_authToken=my_token\n" \
-                       "always-auth = true\n\n" \
-                       "//npm.fury.io/dependabot/:_authToken=my_token")
+                       "//npm.fury.io/dependabot/:_authToken=my_token\n" \
+                       "always-auth = true\n")
             end
 
             context "that uses environment variables everywhere" do
@@ -360,9 +358,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                          "platform-npm/:always-auth=true\n" \
                          "always-auth = true\n" \
                          "registry = https://npm.fury.io/dependabot\n" \
-                         "_authToken=my_token\n" \
-                         "always-auth = true\n\n" \
-                         "//npm.fury.io/dependabot/:_authToken=my_token")
+                         "//npm.fury.io/dependabot/:_authToken=my_token\n" \
+                         "always-auth = true\n")
               end
             end
           end
@@ -383,9 +380,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
               it "adds a global registry line based on the lockfile details" do
                 expect(npmrc_content)
                   .to eq("registry = https://npm.fury.io/dependabot\n" \
-                         "_authToken=my_token\n" \
-                         "always-auth = true\n" \
-                         "//npm.fury.io/dependabot/:_authToken=my_token")
+                         "//npm.fury.io/dependabot/:_authToken=my_token\n" \
+                         "always-auth = true")
               end
             end
           end
@@ -632,9 +628,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
           it "adds a global registry line, and token auth details" do
             expect(npmrc_content)
               .to eq("registry = https://npm.fury.io/dependabot\n" \
-                     "_authToken=my_token\n" \
-                     "always-auth = true\n" \
-                     "//npm.fury.io/dependabot/:_authToken=my_token")
+                     "//npm.fury.io/dependabot/:_authToken=my_token\n" \
+                     "always-auth = true")
           end
 
           context "with basic auth credentials" do
@@ -654,9 +649,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
             it "adds a global registry line, and Basic auth details" do
               expect(npmrc_content)
                 .to eq("registry = https://npm.fury.io/dependabot\n" \
-                       "_auth=c2VjcmV0OnRva2Vu\n" \
-                       "always-auth = true\n" \
-                       "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu")
+                       "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu\n" \
+                       "always-auth = true")
             end
           end
 
@@ -669,9 +663,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                        "strict-ssl = true\n" \
                        "//npm.fury.io/dependabot/:_authToken=secret_token\n" \
                        "registry = https://npm.fury.io/dependabot\n" \
-                       "_authToken=my_token\n" \
-                       "always-auth = true\n\n" \
-                       "//npm.fury.io/dependabot/:_authToken=my_token")
+                       "//npm.fury.io/dependabot/:_authToken=my_token\n" \
+                       "always-auth = true\n")
             end
 
             context "with basic auth credentials" do
@@ -694,9 +687,8 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                          "strict-ssl = true\n" \
                          "//npm.fury.io/dependabot/:_authToken=secret_token\n" \
                          "registry = https://npm.fury.io/dependabot\n" \
-                         "_auth=c2VjcmV0OnRva2Vu\n" \
-                         "always-auth = true\n\n" \
-                         "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu")
+                         "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu\n" \
+                         "always-auth = true\n")
               end
             end
           end
@@ -871,10 +863,9 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
           expect(npmrc_content)
             .to eq(<<~NPMRC.chomp)
               registry = https://npm.pkg.github.com
-              _authToken=my_token
+              //npm.pkg.github.com/:_authToken=my_token
               always-auth = true
               @dsp-testing:registry=https://npm.pkg.github.com
-              //npm.pkg.github.com/:_authToken=my_token
             NPMRC
         end
       end

--- a/npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npmrc_builder_spec.rb
+++ b/npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npmrc_builder_spec.rb
@@ -656,7 +656,6 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                 .to eq("registry = https://npm.fury.io/dependabot\n" \
                        "_auth=c2VjcmV0OnRva2Vu\n" \
                        "always-auth = true\n" \
-                       "always-auth = true\n" \
                        "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu")
             end
           end
@@ -697,7 +696,6 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                          "registry = https://npm.fury.io/dependabot\n" \
                          "_auth=c2VjcmV0OnRva2Vu\n" \
                          "always-auth = true\n\n" \
-                         "always-auth = true\n" \
                          "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu")
               end
             end

--- a/npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npmrc_builder_spec.rb
+++ b/npm_and_yarn/spec/dependabot/npm_and_yarn/file_updater/npmrc_builder_spec.rb
@@ -332,7 +332,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
           it "adds a global registry line, and auth details" do
             expect(npmrc_content)
               .to eq("registry = https://npm.fury.io/dependabot\n" \
-                     "_authToken = my_token\n" \
+                     "_authToken=my_token\n" \
                      "always-auth = true\n" \
                      "//npm.fury.io/dependabot/:_authToken=my_token")
           end
@@ -346,7 +346,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                        "strict-ssl = true\n" \
                        "//npm.fury.io/dependabot/:_authToken=secret_token\n" \
                        "registry = https://npm.fury.io/dependabot\n" \
-                       "_authToken = my_token\n" \
+                       "_authToken=my_token\n" \
                        "always-auth = true\n\n" \
                        "//npm.fury.io/dependabot/:_authToken=my_token")
             end
@@ -360,7 +360,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                          "platform-npm/:always-auth=true\n" \
                          "always-auth = true\n" \
                          "registry = https://npm.fury.io/dependabot\n" \
-                         "_authToken = my_token\n" \
+                         "_authToken=my_token\n" \
                          "always-auth = true\n\n" \
                          "//npm.fury.io/dependabot/:_authToken=my_token")
               end
@@ -383,7 +383,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
               it "adds a global registry line based on the lockfile details" do
                 expect(npmrc_content)
                   .to eq("registry = https://npm.fury.io/dependabot\n" \
-                         "_authToken = my_token\n" \
+                         "_authToken=my_token\n" \
                          "always-auth = true\n" \
                          "//npm.fury.io/dependabot/:_authToken=my_token")
               end
@@ -632,7 +632,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
           it "adds a global registry line, and token auth details" do
             expect(npmrc_content)
               .to eq("registry = https://npm.fury.io/dependabot\n" \
-                     "_authToken = my_token\n" \
+                     "_authToken=my_token\n" \
                      "always-auth = true\n" \
                      "//npm.fury.io/dependabot/:_authToken=my_token")
           end
@@ -654,7 +654,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
             it "adds a global registry line, and Basic auth details" do
               expect(npmrc_content)
                 .to eq("registry = https://npm.fury.io/dependabot\n" \
-                       "_auth = c2VjcmV0OnRva2Vu\n" \
+                       "_auth=c2VjcmV0OnRva2Vu\n" \
                        "always-auth = true\n" \
                        "always-auth = true\n" \
                        "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu")
@@ -670,7 +670,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                        "strict-ssl = true\n" \
                        "//npm.fury.io/dependabot/:_authToken=secret_token\n" \
                        "registry = https://npm.fury.io/dependabot\n" \
-                       "_authToken = my_token\n" \
+                       "_authToken=my_token\n" \
                        "always-auth = true\n\n" \
                        "//npm.fury.io/dependabot/:_authToken=my_token")
             end
@@ -695,7 +695,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
                          "strict-ssl = true\n" \
                          "//npm.fury.io/dependabot/:_authToken=secret_token\n" \
                          "registry = https://npm.fury.io/dependabot\n" \
-                         "_auth = c2VjcmV0OnRva2Vu\n" \
+                         "_auth=c2VjcmV0OnRva2Vu\n" \
                          "always-auth = true\n\n" \
                          "always-auth = true\n" \
                          "//npm.fury.io/dependabot/:_auth=c2VjcmV0OnRva2Vu")
@@ -873,7 +873,7 @@ RSpec.describe Dependabot::NpmAndYarn::FileUpdater::NpmrcBuilder do
           expect(npmrc_content)
             .to eq(<<~NPMRC.chomp)
               registry = https://npm.pkg.github.com
-              _authToken = my_token
+              _authToken=my_token
               always-auth = true
               @dsp-testing:registry=https://npm.pkg.github.com
               //npm.pkg.github.com/:_authToken=my_token

